clippy_utils/src/diagnostics.rs

```diff
@@ -213,6 +213,82 @@ pub fn span_lint_and_sugg<'a, T: LintContext>(
     });
 }
 
+/// Like `span_lint_and_sugg` but with either span or multispan suggestion depending on the number
+/// of its lines.
+///
+/// If the given `sugg` string has more number of lines than
+/// `rustc_errors::emitter::MAX_SUGGESTION_HIGHLIGHT_LINES`, this function will "separate" the
+/// suggestion and span for both top and bottom edge of the code, which CANNOT be auto-fixed,
+/// otherwise this will act like `span_lint_and_sugg` with a combined help message.
+///
+/// # Example
+///
+/// ```text
+/// error: called `map(..).flatten()` on `Option`
+///   --> $DIR/map_flatten.rs:8:10
+///    |
+/// LL |           .map(|x| {
+///    |  __________^
+/// LL | |             if x <= 5 {
+/// LL | |                 Some(x)
+/// LL | |             } else {
+/// ...  |
+/// LL | |         })
+/// LL | |         .flatten();
+///    | |__________________^
+///    |
+///   = note: `-D clippy::map-flatten` implied by `-D warnings`
+/// help: try replacing `map` with `and_then`
+///    |
+/// LL ~         .and_then(|x| {
+/// LL +             if x <= 5 {
+/// LL +                 Some(x)
+///    |
+/// help: and remove the `.flatten()`
+///    |
+/// LL +                 None
+/// LL +             }
+/// LL ~         });
+///    |
+/// ```
+pub fn span_lint_and_sugg_for_edges(
+    cx: &LateContext<'_>,
```

Comment:
Is this alright? the function name, description, content, etc. I'm terrible at phrasing🤣 so if the comment needs improvement you can help propose some changes and I'll directly patch them~

Thank you again~